Teach drag start behaviors to DragGestureRecognizer

edman opened the pull request:
DragGestureRecognizer used to always pass the down point that was initially touched. With this change its behavior become configurable, and onStart can be given either the initial down point, or the point touched at the time onStart is invoked.

This currently doesn't have platform-specific logic since I am only aware of the behavior in Android (from issue #18249).

cc @Hixie

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -163,12 +179,15 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
       _state = _DragState.accepted;
       final Offset delta = _pendingDragOffset;
       final Duration timestamp = _lastPendingEventTimestamp;
+      final Offset startPosition = dragStartBehavior == DragStartBehavior.start
```
please use a switch rather than `?:`. The analyzer will then tell us if we ever add a value to the enum.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -48,7 +57,14 @@ typedef void GestureDragCancelCallback();
+  DragGestureRecognizer({
+    DragStartBehavior dragStartBehavior,
+    Object debugOwner,
+  })  : dragStartBehavior = dragStartBehavior ?? DragStartBehavior.start,
```
defaults should be in the constructor arguments, then you can just assert != null

edman replied:
The problem is the subclasses calling this constructor are always passing a (possibly null) value, so the default constructor argument is never picked up. I could add the default in each of the three subclasses, just thought this was a better option.

Hixie replied:
Yeah, just update all the constructors.

Hixie commented:
The way this is described and implemented, it sounds like you only change the initial delta in the start event, but doesn't this also affect the update events? I would have expected to see _initialPosition get affected in some way. I could be wrong though.

edman commented:
You're very right, it would have been a bug to not clear off the delta after reporting it to onStart. I've updated the code to fix that.

This breaks some tests, I will update them and add some more tomorrow.

edman commented:
Please take a look, tests updated.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -162,20 +178,31 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
-          primaryDelta: _getPrimaryValueFromOffset(delta),
+          delta: _getDeltaForDetails(updateDelta),
+          primaryDelta: _getPrimaryValueFromOffset(updateDelta),
           globalPosition: _initialPosition,
```
is it right that this sends the _initialPosition and not the position sent in onStart?

edman replied:
Updated. But in this line _initialPosition is the same as startPosition.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -162,20 +178,31 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
           sourceTimeStamp: timestamp,
-          delta: _getDeltaForDetails(delta),
-          primaryDelta: _getPrimaryValueFromOffset(delta),
+          delta: _getDeltaForDetails(updateDelta),
```
i'm confused. It seems like if dragStartBehavior is DragStartBehavior.start, the updateDelta will always be zero?

edman replied:
Yes. Because in DragStartBehavior.start, the drag event is considered to begin at the current position (as of when onStart is called).

So `startPosition` consumes any delta that may exist at that moment, after which there is no reason to call onUpdate.

Hixie commented:
I'm surprised this didn't break more tests. Surely it should have broken every test where we look at how far you've dragged when there's something you can tap as well.

edman commented:
It broke only tests that checked that initial behavior (at the time `acceptGesture` is called). Subsequent updates are unaffected.

Hixie commented:
@edman I don't understand why subsequent updates are unaffected. If we used to move by 10 pixels, then 1 pixel, then 1 pixel. with this patch we should now move 0 pixels, then 1 pixel, then 1 pixel, right? The difference is 10 pixels, which seems like it should be a big deal, e.g. to anything that can scroll, since the scroll offset will be off by 10 pixels.

edman commented:
@Hixie events will be reported after the gesture recognizer commits, and only after that. 

I.e. say there are two competing recognizers, a vertical drag vs a horizontal drag. When the first event dispatched is a 10px vertical move, the vertical drag will win, its `onStart` will be triggered with the end position of that event. From then on subsequent move events will trigger `onUpdates` normally.

On the other hand if there is only one gesture recognizer, it will commit immediately during the down event, calling `onStart`, after which the 10px move will trigger `onUpdate` normally.

So yes, events that happen before a recognizer commits are discarded. I am not sure how often events are dispatched, how soon a competing recognizer is found winner and commits, and how many pixels would be discarded as result in practice.

My understanding was this was the intended behavior, as the vertical drag in the example above commits immediately after the 10px move, so that becomes the commit point future deltas should be based on. Apparently I am wrong? What are the requirements?

Hixie commented:
What you describe sounds correct, I'm just surprised it didn't end up in more tests having issues.

Suppose you have two gesture recognizers, one of which is a vertical drag. You drag 10 pixels, and it triggers starting there. You then drag one more pixel, so the next update should be for 1 px, then you drag another, then next is another 1 px. Total distance: 2 pixels. So if you have a test with a drag gesture recognizer, it should now drag 10 pixels less than before (where we would have sent updated 10 px, 1 px, 1px instead of 0 px, 1 px, 1 px).

edman commented:
Yes you're right. Not many tests were affected because only two have competing recognizers. Also tests do checks on a per-event basis, meaning in your example they wouldn't check the total delta was 12px, they would check after each event for deltas of 10px, then 1px, then 1px. So updating the first check is enough.

- In "should recognize pan" I changed the first updated delta check to null.
https://github.com/flutter/flutter/blob/ac3f542ab2bb9003a12c3fd7765eb389324e1177/packages/flutter/test/gestures/drag_test.dart#L60

- In "drag with multiple pointers" I made it use the old down behavior.
https://github.com/flutter/flutter/blob/ac3f542ab2bb9003a12c3fd7765eb389324e1177/packages/flutter/test/gestures/drag_test.dart#L250-L254

Other than that, I added tests to cover the differences in those initial events for each behavior.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -162,21 +180,32 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
-          globalPosition: _initialPosition,
+          delta: _getDeltaForDetails(updateDelta),
+          primaryDelta: _getPrimaryValueFromOffset(updateDelta),
+          globalPosition: startPosition,
```
we should probably be adding updateDelta to startPosition here

Hixie commented:
Looks like some tests are failing now?

Hixie commented:
@edman Will you be able to continue working on this? It's totally fine if not; we can take it over (though it might take longer). I'm just asking because I'm trying to clean out our review queue.

edman commented:
@Hixie yes I can still take this. Some tests will have to be updated because of the changes in #21138. Will do so tonight.

edman committed:
Fix globalPosition inconsistency in onUpdates

edman committed:
Teach drag start behaviors to DragGestureRecognizer

DragGestureRecognizer used to always pass the down point that was
initially touched. With this change its behavior become configurable,
and onStart can be given either the initial down point, or the point
touched at the time onStart is invoked. These two points can be
different when there are competing recognizers.

edman committed:
Update existing tests in drag_test.dart with drag behavior

edman commented:
There's some inconsistency to how the `globalPosition` is reported to `onUpdate` callbacks in `acceptGesture` vs `handleEvent` after #21138. 

In `acceptGesture` the position reported is offset by the primary delta, while in `handleEvent` the position is offset by the full delta.

I added an initial commit to fix that by offsetting the full delta in both cases and improved tests to catch this distinction. Please have a look.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -48,7 +59,14 @@ typedef GestureDragCancelCallback = void Function();
+  })  : assert(dragStartBehavior != null),
+        super(debugOwner: debugOwner);
+
+  /// Configure the behavior of offsets sent to [onStart]
```
nit: this is really trivial but please end sentences in documentation with a period.

Hixie commented on packages/flutter/lib/src/gestures/monodrag.dart:
```diff
@@ -163,20 +181,32 @@ abstract class DragGestureRecognizer extends OneSequenceGestureRecognizer {
-          globalPosition: _initialPosition + delta,
+          delta: updateDelta,
+          primaryDelta: _getPrimaryValueFromOffset(updateDelta),
+          globalPosition: startPosition + delta,
```
Won't this cause delta to be added twice in some cases? Specifically in the DragStartBehavior.start case, since we add delta to initialPosition to get startPosition?

jslavitz replied:
It seems like it might not make sense to send an onUpdate call from this function (acceptGesture), if the dragStartBehaviour is set to `start` because the delta is going into the onStart call.

edman replied:
This code is guarded by the `updateDelta != Offset.zero` check, which won't pass for `DragStartBehavior.start` due to the switch logic above.

Result being that `onUpdate` is never called with `start` behavior; and when it is called (in `down` behavior), `startPosition` does not contain the delta, so there's no double counting.

I agree this flow is not the most direct and readable but it seemed best to me as it kept the code simpler. Happy to change if you think there's a better way.

Hixie commented:
Would it make sense to actually update _initialPosition? I'm thinking about what will happen if we ever add a field to the details to describe the commit position. Right now, the naive approach would report _initialPosition, which would then mean it disagrees with startPosition.

edman committed:
Update _initialPosition to match drag behavior

edman commented:
Indeed `_initialPosition` does not reflect the position sent to `onStart`, and once `acceptGesture` ends that information is lost.

Added a commit to update `_initialPosition` with the actual commit position, as what is reported to `onStart`.
